lib/ansible/executor/task_executor.py

```diff
@@ -1124,14 +1125,19 @@ def _get_action_handler_with_module_context(self, templar):
         module = self._shared_loader_obj.module_loader.find_plugin_with_context(
             self._task.action, collection_list=collections
         )
-        if not module.resolved or not module.action_plugin:
+
+        if not module.resolved:
+            # if we could not resolve, we return a None context, but handler will still be set
             module = None
-        if module is not None:
+
+        if module is not None and module.action_plugin:
```

Comment:
This can be reverted with the last commit, it's the cause of the failure for the `legacy does not use builtin defaults` test. It's a subtle change but if a module exists but hasn't defined an `action_plugin` in its runtime.yml, this allows it to take precedence over an action plugin by the same name (which might resolve to a different name).